### PR TITLE
Import-util: URL processing tweaks

### PR DESCRIPTION
**Pull request by poettering**
This cleans up and modernizes our URL handling, and adds a test case for it.

### CODE REVIEW

File: src/test/test-import-util.c [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+static void test_import_url_last_component_one(const char *input, const char *output, int ret) {
+        _cleanup_free_ char *s = NULL;
+
+        assert_se(import_url_last_component(input, &s) == ret);
+        assert_se(streq_ptr(output, s));
+}
+
+static void test_import_url_last_component(void) {
+        test_import_url_last_component_one("https://foobar/waldo/quux", "quux", 0);
+        test_import_url_last_component_one("https://foobar/waldo/quux", "quux", 0);
```
Comment (yuwata) [inline src/test/test-import-util.c]: The test is duplicated.

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/shared/import-util.c [hunk 2]
```diff
@@ -14,53 +14,114 @@
...
+        e = h + strcspn(h, "?#"); /* Cut off "Query" and "Fragment" */
 
-        while (e > url && e[-1] != '/')
+        while (e > h && e[-1] == '/') /* Eat trailing slashes */
                 e--;
 
-        if (e <= url)
-                return -EINVAL;
+        /* Drop the specified number of components from the end. Note that this is pretty lenient: if there
+         * are less component we silently drop those and then append the suffix to the top. */
```
Comment (keszybz) [inline src/shared/import-util.c:109]: less component we silently drop → fewer components, we just drop
to the top → ''

File: src/shared/import-util.c [hunk 3]
```diff
@@ -14,53 +14,114 @@
...
+        const char *d;
+        size_t n;
+
+        /* A very very lenient implementation of RFC3986 Section 3.2 */
+
+        /* Find colon separating protocol and hostname */
+        d = strchr(url, ':');
+        if (!d || url == d)
+                return NULL;
+        d++;
```
Comment (keszybz) [inline src/shared/import-util.c:27]: Strictly speaking, RFC 3986 says that the hostname ("authority") may be empty:

   The authority component is preceded by a double slash ("//") and is
   terminated by the next slash ("/"), question mark ("?"), or number
   sign ("#") character, or by the end of the URI.

So this parsers eats too much in some cases, and not enough in others. I'm not sure if it matters for the cases where we are using this... But maybe it'd be better to make the parser follow the RFC if it's not much more work?
```C
d = url + strspn(url, "0…9A…Za-z+-.");
if (d == url)
         return NULL;
d = startswith(d, "://");
if (!d)
         return NULL;
return d + strcpsn(d, "/?#");
```
?

Comment (poettering) [reply src/shared/import-util.c:27]: I wanted to be extra lenient here, i.e. not make restrictions on the charset and stuff. i.e. look for the chars that matter, but ignore all others. 

Your proposed code looks for `://` but many urls don't have them, such as mailto and stuff. i'd be careful with that...

The function makes clear that we skip the hostname, i.e. expect there is one, which is an OK assumption for the protocols we care abou there.

**Pull request merged**